docs/book/provider_implementations/populate_controllers.md

```diff
@@ -0,0 +1,159 @@
+# Controllers and Reconciliation
+
+From the [kubebuilder book][controller]:
+
+> Controllers are the core of Kubernetes, and of any operator.
+>
+> It’s a controller’s job to ensure that, for any given object, the actual state of the world (both the cluster state, and potentially external state like running containers for Kubelet or loadbalancers for a cloud provider) matches the desired state in the object. 
+> Each controller focuses on one root Kind, but may interact with other Kinds.
+>
+> We call this process reconciling.
+
+[controller]: https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller
+
+Right now, we can create objects in our API but we won't do anything about it. Let's fix that. 
+
+# Let's see the Code
+
+Kubebuilder has created our first controller in `controllers/mailguncluster_controller.go`. Let's take a look at what got generated:
+
+```go
+// MailgunClusterReconciler reconciles a MailgunCluster object
+type MailgunClusterReconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=mailgunclusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=mailgunclusters/status,verbs=get;update;patch
+
+func (r *MailgunClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("mailguncluster", req.NamespacedName)
+
+	// your logic here
+
+	return ctrl.Result{}, nil
+}
+```
+
+## State
+
+Let's focus on that `struct` first. 
+First, a word of warning: no guarantees are made about parallel access, both on one machine or multiple machines.
+That means you should not store any important state in memory: if you need it, write it into a Kubernetes object and store it.
```

Comment:
I think there are a few concepts here:

1. You should think about whether or not you want to allow multiple MailgunClusters to be reconciled at the same time by the same manager process. The default is to run a single worker, but you could choose to increase this, optionally exposing it to the user to control. If you do, you need to make sure you code defensively if multiple goroutines need to modify the same field on `MailgunClusterReconciler` at the same time (i.e., use a mutex).
1. You should think about whether or not you want to allow multiple instances of your manager process to be active at the same time. The default for kubebuilder projects is yes, which means you could have two managers running at the same time, both trying to reconcile the same `MailgunCluster`. If you want to avoid this, and limit to just a single active manager at a time, you'll need to enable leader election in your manager.
1. If you generate any important state, and it's not possible to reconstruct it (or maybe it's expensive to calculate), you should consider storing it somewhere (as a Kubernetes object, or whatnot).